tests/cpgrid/avoidNNCinLGRsCpGrid_test.cpp

```diff
@@ -0,0 +1,169 @@
+// $Date$
+//
+// $Revision$
+//
+//===========================================================================
+
+/*
+  Copyright 2023 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+
+#define NVERBOSE
+
+#define BOOST_TEST_MODULE CpGridWithNNC
+
+#include <boost/test/unit_test.hpp>
+
+#include <dune/common/version.hh>
+#include <dune/grid/common/mcmgmapper.hh>
+#include <opm/input/eclipse/Deck/Deck.hpp>
+#include <opm/input/eclipse/Parser/Parser.hpp>
+#include <opm/input/eclipse/EclipseState/EclipseState.hpp>
+#include <opm/grid/CpGrid.hpp>
+
+#include <iostream>
+#include <vector>
+#include <utility>
+
+struct Fixture
+{
+    Fixture()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        Dune::MPIHelper::instance(m_argc, m_argv);
+        Opm::OpmLog::setupSimpleDefaultLogging();
+    }
+
+    static int rank()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        return Dune::MPIHelper::instance(m_argc, m_argv).rank();
+    }
+};
+
+void testCase(const std::string& deckString,
+              const Opm::NNC& nnc,
+              const std::vector<std::array<int,3>>& cells_per_dim_vec,
+              const std::vector<std::array<int,3>>& startIJK_vec,
+              const std::vector<std::array<int,3>>& endIJK_vec,
+              const std::vector<std::string>& lgr_name_vec)
+{
+
+    Opm::Parser parser;
+    const auto deck = parser.parseString(deckString);
+    Opm::EclipseState es(deck);
+    es.appendInputNNC(nnc.input());
+
+    Dune::CpGrid grid;
+    grid.processEclipseFormat(&es.getInputGrid(), &es, false, false, false);
+
+    const auto& data = grid.data_;
+    const auto& all_lgrs_cells = (*data[0]).getPatchesCells(startIJK_vec, endIJK_vec);
+    const auto& hasNNC =  (*data[0]).hasNNCs(all_lgrs_cells);
+    std::cout << std::boolalpha;
+    std::cout << hasNNC << '\n';
+
+    if(hasNNC){
+        BOOST_CHECK_THROW(grid.addLgrsUpdateLeafView(cells_per_dim_vec, startIJK_vec, endIJK_vec, lgr_name_vec), std::logic_error);
+    }
+    else{
+        grid.addLgrsUpdateLeafView(cells_per_dim_vec, startIJK_vec, endIJK_vec, lgr_name_vec);
+    }
```

Comment:
I am a bit skeptical about this approach:

We are using hasNNcs(...) to decide what to test, but hasNNcs(...) is called inside addLgrsUpdateLeafView to decide whether to throw. Seems like we do not really check anything. 

I think we should rather for each test know when whether or not there is an NNC to one of the cells that will be refined and pass another argument (e.g. hasNNCs) to the testCase function.

Maybe we should also add a testCase that won't throw in addition if there is none, yet.

Reply:
Yes! The reason of the if-statement is due to having 2 cases where it should throw and, the last one, that has an NNC, but on cells not contain in any LGR. We can also remove that case, or re-design the test without the if-statement and adding an extra argument in the function, too.

Reply:
What I meant is that this decision should not be based on calling the hasNNC methods, but there should be a parameter that tells the test whether this should throw or not.

Then for e.g. NNCoutsideLgrs this parameter would be false:
```
testCase(deckString, nnc,  cells_per_dim_vec, startIJK_vec, endIJK_vec, lgr_name_vec, false);
```
In the current version if hasNNCs is wrong (returns true if there are none) this will not be detected.